### PR TITLE
[ntuple] Fix read request coalescing when opening file from anchor

### PR DESCRIPTION
**Pull request by jblomer**
When we open an RNTuple from an anchor, we don't read the anchor with the mini file reader and therefore we must initialize the max key size manually from the given anchor. Otherwise, page coalescing won't work.

@Dr15Jones FYI.

### CODE REVIEW

File: tree/ntuple/v7/src/RPageStorageFile.cxx [hunk 1]
```diff
@@ -313,8 +313,11 @@ void ROOT::Experimental::Internal::RPageSourceFile::LoadStructureImpl()
 {
    // If we constructed the page source with (ntuple name, path), we need to find the anchor first.
    // Otherwise, the page source was created by OpenFromAnchor()
-   if (!fAnchor)
+   if (!fAnchor) {
       fAnchor = fReader.GetNTuple(fNTupleName).Unwrap();
+   } else {
+      fReader.SetMaxKeySize(fAnchor->GetMaxKeySize());
+   }
```
Comment (pcanal) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:320]: I am a bit confused.  Why is it not:
```suggestion
   if (!fAnchor) {
      fAnchor = fReader.GetNTuple(fNTupleName).Unwrap();
      fReader.SetMaxKeySize(fAnchor->GetMaxKeySize());
   }
```
and then the `SetMaxKeySize` also on line 305 ?

Comment (jblomer) [reply tree/ntuple/v7/src/RPageStorageFile.cxx:320]: The mini file reader will set the max key size as a side effect of reading the anchor.
If we already have the anchor, we don't need to read it again.

**Comment by Dr15Jones** [comment]
So this affected reading an RNTuple back from a TFile? What would be the visible change in behavior after this fix?

File: tree/ntuple/v7/src/RPageStorageFile.cxx (hunk 1 above)
Comment (jblomer) [reply tree/ntuple/v7/src/RPageStorageFile.cxx:320]: @silverweed What do you think? I agree that the current situation is not ideal.

**Comment by jblomer** [comment]
> So this affected reading an RNTuple back from a TFile? What would be the visible change in behavior after this fix?

The effect is that ROOT uses much fewer read requests to get the data. At the moment, every page is a read request. When the fix is merged, I expect more or less $(number of top-level fields + number of clusters)$.

File: tree/ntuple/v7/src/RPageStorageFile.cxx (hunk 1 above)
Comment (jblomer) [reply tree/ntuple/v7/src/RPageStorageFile.cxx]: I added @pcanal's suggestion to the PR

File: tree/ntuple/v7/test/ntuple_view.cxx [hunk 2]
```diff
@@ -271,3 +273,72 @@ TEST(RNTuple, ViewStandardIntegerTypes)
...
+
+   auto ntpl = std::unique_ptr<RNTuple>(file.Get<RNTuple>("ntpl"));
+   auto reader = RNTupleReader::Open(*ntpl);
+   reader->EnableMetrics();
+
+   std::optional<ROOT::Experimental::RNTupleView<void>> viewPx;
+   std::optional<ROOT::Experimental::RNTupleView<void>> viewPy;
+   std::optional<ROOT::Experimental::RNTupleView<void>> viewPz;
+
+   float px, py, pz;
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_view.cxx:310]: maybe initialize these variables to silence the (spurious) compiler warnings...

File: tree/ntuple/v7/test/ntuple_view.cxx [hunk 3]
```diff
@@ -271,3 +273,72 @@ TEST(RNTuple, ViewStandardIntegerTypes)
...
+
+TEST(RNTuple, ViewFrameworkUse)
+{
+   TMemFile file("memfile.root", "RECREATE");
+
+   {
+      auto model = RNTupleModel::Create();
+      auto ptrPx = model->MakeField<float>("px");
+      auto ptrPy = model->MakeField<float>("py");
+      model->MakeField<bool>("trigger");
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_view.cxx:285]: What is `trigger` used for?

File: tree/ntuple/v7/test/ntuple_view.cxx [hunk 4]
```diff
@@ -271,3 +273,72 @@ TEST(RNTuple, ViewStandardIntegerTypes)
...
+            viewPz = reader->GetView<void>("pz", &pz);
+         }
+         (*viewPz)(i);
+         EXPECT_FLOAT_EQ(0.4 + i, pz);
+      }
+   }
+
+   // Ensure that cluster prefetching and smearing of read requests works
+   EXPECT_LT(reader->GetDescriptor().GetNClusters(),
+             reader->GetMetrics().GetCounter("RNTupleReader.RPageSourceFile.nClusterLoaded")->GetValueAsInt());
```
Comment (hahnjo) [inline tree/ntuple/v7/test/ntuple_view.cxx:339]: Why do we expect to load more clusters than there are in the file?

**Review by hahnjo — APPROVED** [review]
LGTM as a fix, other improvements can be done in future PRs. Some comments on the added test inline.

**Pull request merged**